forbidigo/forbidigo.go

```diff
@@ -163,15 +166,36 @@ func (v *visitor) Visit(node ast.Node) ast.Visitor {
 			return nil
 		}
 		return v
+	// The following two are handled below.
 	case *ast.SelectorExpr:
 	case *ast.Ident:
+	// Everything else isn't.
 	default:
 		return v
 	}
+
+	// The text as it appears in the source is always used because issues
+	// use that. The other texts to match against are extracted when needed
+	// by a pattern. They are nil when not evaluated yet and point to
+	// the empty string when there is nothing to match against.
+	srcText := v.textFor(node)
+	var pkgText *string
 	for _, p := range v.linter.patterns {
-		if p.pattern.MatchString(v.textFor(node)) && !v.permit(node) {
+		if p.matchWithPackage && pkgText == nil {
+			t := v.pkgTextFor(node)
+			pkgText = &t
+		}
+
+		matchText := ""
+		switch {
+		case p.matchWithPackage:
+			matchText = *pkgText
+		default:
+			matchText = srcText
+		}
```

Comment:
switch seems a little verbose here.  Maybe just this:
```suggestion
		matchText =: srcText
		if p.matchWithPackage {
			if pkgText == nil {
				t := v.pkgTextFor(node)
				pkgText = &t
			}
			matchText := pkgText
		}
```